docs/machine-learning/basic-concepts-model-training-in-mldotnet.MD

```diff
@@ -0,0 +1,169 @@
+}
+```
+
+You can easily imagine how *a sequence of estimators can be phrased as an estimator* of its own. In ML.NET, we rely on this property to create 'learning pipelines' that chain together different estimators:
+
+```c#
+var env = new LocalEnvironment(); // Initialize the ML.NET environment.
+var estimator = new ConcatEstimator(env, "Features", "SepalLength", "SepalWidth", "PetalLength", "PetalWidth")
+    .Append(new ToKeyEstimator(env, "Label"))
+    .Append(new SdcaMultiClassTrainer(env, "Features", "Label")) // This is the actual 'machine learning algorithm'.
+    .Append(new ToValueEstimator(env, "PredictedLabel"));
+
+var endToEndModel = estimator.Fit(data); // This now contains all the transformers that were used at training.
+```
+
+One important property of estimators is that *estimators are eager, not lazy*: every call to `Fit` is causing 'learning' to happen, which is potentially a time-consuming operation.
+
+## Prediction function
+
+The prediction function can be seen as a machine that applies a transformer to one 'row', such as at prediction time.
+
+Once we obtain the model (a *transformer* either trained via `Fit()`, or loaded from somewhere), we can use it to make 'predictions' using the normal calls to `model.Transform(data)`. However, when we use this model in a real life scenario, we often don't have a whole 'batch' of examples to predict on. Instead, we have one example at a time, and we need to make timely predictions on them immediately.
+
+Of course, we can reduce this to the batch prediction:
+
+- Create a data view with exactly one row.
+- Call `model.Transform(data)` to obtain the 'predicted data view'.
+- Get a cursor over the resulting data.
+- Advance the cursor one step to get to the first (and only) row.
+- Extract the predicted values out of it.
+
+The above algorithm can be implemented using the [schema comprehension](https://github.com/dotnet/machinelearning/blob/master/docs/code/SchemaComprehension.md), with two user-defined objects `InputExample` and `OutputPrediction` as follows:
+
+```c#
+var inputData = env.CreateDataView(new InputExample[] { example });
+var outputData = model.Transform(inputData);
+var output = outputData.AsDynamic.AsEnumerable<OutputPrediction>(env, reuseRowObject: false).Single();
+```
+
+But this would be cumbersome, and would incur performance costs. 
+Instead, we have a 'prediction function' object that performs the same work, but faster and more convenient, via an extension method `MakePredictionFunction`:
+
+```c#
+var predictionFunc = model.MakePredictionFunction<InputExample, OutputPrediction>(env);
+var output = predictionFunc.Predict(example);
+```
+
+The same `predictionFunc` can (and should!) be used multiple times, thus amortizing the initial cost of `MakePredictionFunction` call.
+
+The prediction function is *not re-entrant / thread-safe*: if you want to conduct predictions simultaneously with multiple threads, you need to have a prediction function per thread.
```

Comment:
This really confuses me.  Suppose I have a scenario where I have an ASP.NET MVC Web API endpoint that performs a prediction. The return of `MakePredictionFunction` cannot (or should not) be shared across threads. OK, so I should make a new one on each web request, right? Well, that sounds like it is a lot of work to perform my predictions. Wouldn't that hurt my scale out by doing this expensive operation on each request? Or, if I serialize all access to the predictionFunc, I can't scale out well. Something seems missing in this explanation.

Reply:
Deadline constraints; will be handled later.

Reply:
CC: @GalOshri @CESARDELATORRE @asthana86 @BillWagner  - Going to remove for now and discuss offline.